payments-core/src/main/java/com/stripe/android/view/CardBrandViewRebuild.kt

```diff
@@ -0,0 +1,249 @@
+    }
+
+    override fun onRestoreInstanceState(state: Parcelable?) {
+        val savedState = state as? SavedState
+        this.state = savedState?.state ?: State()
+        determineCardBrandToDisplay(
+            this.state.userSelectedBrand,
+            this.state.brand,
+            this.state.possibleBrands,
+            this.state.merchantPreferredNetworks
+        )
+        updateBrandSpinner()
+        super.onRestoreInstanceState(savedState?.superState)
+    }
+
+    @Parcelize
+    internal data class SavedState(
+        val superSavedState: Parcelable?,
+        val state: State,
+    ) : BaseSavedState(superSavedState), Parcelable
+
+    private fun updateBrandSpinner() {
+        chevron.visibility = if (possibleBrands.size > 1)  View.VISIBLE else View.GONE
+        if (possibleBrands.size > 1) {
+            val adapter = BrandAdapter(context, possibleBrands)
+            this.setOnClickListener { listPopup.show() }
+            listPopup.setAdapter(adapter)
+
+            listPopup.setOnItemClickListener { parent, view, position, id ->
+                possibleBrands.getOrNull(position - 1)?.let {
+                    iconView.setBackgroundResource(it.icon)
+                    handleBrandSelected(it)
+                }
+                listPopup.dismiss()
+            }
+            listPopup.anchorView = iconView
+            listPopup.width = measureContentWidth(adapter)
+            listPopup.setSelection(possibleBrands.indexOf(brand))
+        } else {
+            this.setOnClickListener(null)
+        }
+    }
+
+    private fun measureContentWidth(adapter: BrandAdapter): Int {
+        val widthMeasureSpec = MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)
+        val heightMeasureSpec = MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)
+        val count = adapter.count
+        var maxWidth = 0
+
+        for (i in 0..<count) {
+            val itemView = adapter.getView(i, null, this)
+            itemView.measure(widthMeasureSpec, heightMeasureSpec)
+            maxWidth = maxWidth.coerceAtLeast(itemView.measuredWidth)
+        }
+
+        return maxWidth
+    }
+}
+
+class BrandAdapter(
+    context: Context,
+    private val brands: List<CardBrand?>,
+) : ArrayAdapter<CardBrand?>(context, 0, brands) {
+
+    private val inflater = LayoutInflater.from(context)
+
+    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+        val view = if (position == 0) {
+            inflater.inflate(R.layout.stripe_select_card_brand_view, parent, false)
+        } else {
+            inflater.inflate(R.layout.stripe_card_brand_spinner_view, parent, false)
+        }
+
+        if (position > 0) {
+            brands.getOrNull(position - 1)?.let {
+                view.findViewById<ImageView>(R.id.brand_icon)?.setBackgroundResource(it.icon)
+                view.findViewById<TextView>(R.id.brand_text)?.text = it.displayName
+            }
```

Comment:
We also need to highlight the currently selected brand, but I’m sure you’re already working on that.

Reply:
💯